sdk/core/core-https/src/pipeline.ts

```diff
@@ -0,0 +1,285 @@
+  private _policies: PipelineDescriptor[] = [];
+  private _orderedPolicies?: PipelinePolicy[];
+
+  private constructor(policies: PipelineDescriptor[] = []) {
+    this._policies = policies;
+    this._orderedPolicies = undefined;
+  }
+
+  public addPolicy(policy: PipelinePolicy, options: AddPolicyOptions = {}): void {
+    if (options.phase && options.afterPhase) {
+      throw new Error("Policies inside a phase cannot specify afterPhase.");
+    }
+    if (options.phase && !ValidPhaseNames.has(options.phase)) {
+      throw new Error(`Invalid phase name: ${options.phase}`);
+    }
+    if (options.afterPhase && !ValidPhaseNames.has(options.afterPhase)) {
+      throw new Error(`Invalid phase name: ${options.afterPhase}`);
+    }
+    this._policies.push({
+      policy,
+      options
+    });
+    this._orderedPolicies = undefined;
+  }
+
+  public removePolicy(options: { name?: string; phase?: string }): PipelinePolicy[] {
+    const removedPolicies: PipelinePolicy[] = [];
+
+    this._policies = this._policies.filter((policyDescriptor) => {
+      if (
+        (options.name && policyDescriptor.policy.name === options.name) ||
+        (options.phase && policyDescriptor.options.phase === options.phase)
+      ) {
+        removedPolicies.push(policyDescriptor.policy);
+        return false;
+      } else {
+        return true;
+      }
+    });
+    this._orderedPolicies = undefined;
+
+    return removedPolicies;
+  }
+
+  public sendRequest(
+    httpsClient: HttpsClient,
+    request: PipelineRequest
+  ): Promise<PipelineResponse> {
+    const policies = this.getOrderedPolicies();
+
+    const pipeline = policies.reduceRight<SendRequest>(
+      (next, policy) => {
+        return (request: PipelineRequest) => {
+          return policy.sendRequest(request, next);
+        };
+      },
+      (request: PipelineRequest) => httpsClient.sendRequest(request)
+    );
+
+    return pipeline(request);
+  }
+
+  public getOrderedPolicies(): PipelinePolicy[] {
+    if (!this._orderedPolicies) {
+      this._orderedPolicies = this.orderPolicies();
+    }
+    return this._orderedPolicies;
+  }
+
+  public clone(): Pipeline {
+    return new HttpsPipeline(this._policies);
+  }
+
+  public static create(): Pipeline {
+    return new HttpsPipeline();
+  }
+
+  private orderPolicies(): PipelinePolicy[] {
+    const result: PipelinePolicy[] = [];
+
+    const policyMap: Map<string, PolicyGraphNode> = new Map<string, PolicyGraphNode>();
+    const noPhase = new Set<PolicyGraphNode>();
+    const serializePhase = new Set<PolicyGraphNode>();
+    const retryPhase = new Set<PolicyGraphNode>();
+
+    function getPhase(phase: PipelinePhase | undefined): Set<PolicyGraphNode> {
+      if (phase === "Retry") {
+        return retryPhase;
+      } else if (phase === "Serialize") {
+        return serializePhase;
+      } else {
+        return noPhase;
+      }
+    }
+
+    for (const descriptor of this._policies) {
```

Comment:
I used to reach for forEach all the time when it was first introduced, but honestly I prefer for/of now.

for/of avoids a function call, you can break out of the loop early, and it works with any iterable rather than only arrays.